src/main/java/service/impl/ShopServiceImpl.java

```diff
@@ -0,0 +1,39 @@
+package service.impl;
+
+import db.Storage;
+import java.util.List;
+import model.FruitTransaction;
+import service.ShopService;
+import strategy.OperationStrategy;
+
+public class ShopServiceImpl implements ShopService {
+    private OperationStrategy operationStrategy;
+
+    public ShopServiceImpl(OperationStrategy operationStrategy) {
+        this.operationStrategy = operationStrategy;
+    }
+
+    @Override
+    public void process(List<FruitTransaction> transactions) {
+        for (FruitTransaction transaction : transactions) {
+            if (!Storage.reports.containsKey(transaction.getFruit())) {
+                Storage.reports.put(transaction.getFruit(), getHowManyHaveChanged(transaction));
+            } else {
+                Storage.reports.put(transaction.getFruit(), getBalance(transaction)
+                        + getHowManyHaveChanged(transaction));
+            }
+            if (getBalance(transaction) < 0) {
+                throw new RuntimeException("The quantity is not enough to purchase");
+            }
+        }
+    }
```

Comment:
The service class should not contain business logic directly. It is better to delegate the execution of operations to the strategy handlers. Also, the check for a negative balance should be done within the operation handler for the 'purchase' operation, not in the service class.